File: pwiz/data/msdata/MSNumpress.cpp
//
// $Id$
//
//
// Original author: Johan Teleman <johan.teleman@immun.lth.se>
//
// Copyright 2013 Johan Teleman
//
// Licensed under the Apache License, Version 2.0 (the "License"); 
// you may not use this file except in compliance with the License. 
// You may obtain a copy of the License at 
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software 
// distributed under the License is distributed on an "AS IS" BASIS, 
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
// See the License for the specific language governing permissions and 
// limitations under the License.
//

#define PWIZ_SOURCE

#include <cstdio>
#include <stdexcept>
#include <vector>
#include <iostream>
#include <cmath>
#include <algorithm>
#include "MSNumpress.hpp"

namespace pwiz {
namespace msdata {
namespace MSNumpress {

using std::cout;
using std::cerr;
using std::endl;
using std::min;
using std::max;
using std::abs;


const int ONE = 1;
bool is_big_endian() {
    return *((char*)&(ONE)) == 1;
}
bool IS_BIG_ENDIAN = is_big_endian();



/////////////////////////////////////////////////////////////

PWIZ_API_DECL
void encodeFixedPoint(
        double fixedPoint, 
        unsigned char *result
) {
    int i;
    unsigned char *fp = (unsigned char*)&fixedPoint;
    for (i=0; i<8; i++) {
        result[i] = fp[is_big_endian() ? (7-i) : i];
    }
}


PWIZ_API_DECL
double decodeFixedPoint(
        const unsigned char *data
) {
    int i;
    double fixedPoint;
    unsigned char *fp = (unsigned char*)&fixedPoint;
        
    for (i=0; i<8; i++) {
        fp[i] = data[is_big_endian() ? (7-i) : i];
    }
    
    return fixedPoint;
}

/////////////////////////////////////////////////////////////

/**
 * Encodes the int x as a number of halfbytes in res. 
 * res_length is incremented by the number of halfbytes, 
 * which will be 1 <= n <= 9
 */
PWIZ_API_DECL
void encodeInt(
        const int x,
        unsigned char* res,
        size_t *res_length    
) {
    int i, l, m;
    int mask = 0xf0000000;
    unsigned int init = x & mask;

    if (init == 0) {
        l = 8;
        for (i=0; i<8; i++) {
            m = mask >> (4*i);
            if ((x & m) != 0) {
                l = i;
                break;
            }
        }
        res[0] = l;
        for (i=l; i<8; i++) {
            res[1+i-l] = x >> (4*(i-l));
        }
        *res_length += 1+8-l;

    } else if (init == mask) {
        l = 7;
        for (i=0; i<8; i++) {
            m = mask >> (4*i);
            if ((x & m) != m) {
                l = i;
                break;
            }
        }
        res[0] = l + 8;
        for (i=l; i<8; i++) {
            res[1+i-l] = x >> (4*(i-l));
        }
        *res_length += 1+8-l;

    } else {
        res[0] = 0;
        for (i=0; i<8; i++) {
            res[1+i] = x >> (4*i);
        }
        *res_length += 9;

    }
}



/**
 * Decodes an int from the half bytes in bp. Lossless reverse of encodeInt 
 */
PWIZ_API_DECL
void decodeInt(
        const unsigned char *data,
        size_t *di,
        int *half,
        int *res
) {
    size_t n;
    size_t i;
    int mask, m;
    unsigned char head;
    unsigned char hb;

    if (*half == 0) {
        head = data[*di] >> 4;
    } else {
        head = data[*di] & 0xf;
        (*di)++;
    }

    *half = 1-(*half);
    *res = 0;
    
    if (head <= 8) {
        n = head;
    } else { // leading ones, fill n half bytes in res
        n = head - 8;
        mask = 0xf0000000;
        for (i=0; i<n; i++) {
            m = mask >> (4*i);
            *res = *res | m;
        }
    }
    
    if (n == 8) {
        return;
    }
    
    for (i=n; i<8; i++) {
        if (*half == 0) {
            hb = data[*di] >> 4;
        } else {
            hb = data[*di] & 0xf;
            (*di)++;
        }
        *res = *res | (hb << ((i-n)*4));
        *half = 1 - (*half);
    }
}




/////////////////////////////////////////////////////////////

PWIZ_API_DECL
double optimalLinearFixedPointMass(
		const double *data, 
		size_t dataSize,
        double mass_acc
) {
	if (dataSize < 3) return 0; // we just encode the first two points as floats

    // We calculate the maximal fixedPoint we need to achieve a specific mass
    // accuracy. Note that the maximal error we will make by encoding as int is
    // 0.5 due to rounding errors.
    double maxFP = 0.5 / mass_acc;

    // There is a maximal value for the FP given by the int length (32bit)
    // which means we cannot choose a value higher than that. In case we cannot
    // achieve the desired accuracy, return failure (-1).
    double maxFP_overflow = optimalLinearFixedPoint(data, dataSize);
    if (maxFP > maxFP_overflow) return -1;

    return maxFP;
}

PWIZ_API_DECL
double optimalLinearFixedPoint(
        const double *data, 
        size_t dataSize
) {
    /*
     * safer impl - apparently not needed though
     *
    if (dataSize == 0) return 0;
    
    double maxDouble = 0;
    double x;

    for (size_t i=0; i<dataSize; i++) {
        x = data[i];
        maxDouble = max(maxDouble, x);
    }

    return floor(0xFFFFFFFF / maxDouble);
    */
    if (dataSize == 0) return 0;
    if (dataSize == 1) return floor(0x7FFFFFFFl / data[0]);
    double maxDouble = max(data[0], data[1]);
    double extrapol;
    double diff;

    for (size_t i=2; i<dataSize; i++) {
        extrapol = data[i-1] + (data[i-1] - data[i-2]);
        diff = data[i] - extrapol;
        maxDouble = max(maxDouble, ceil(abs(diff)+1));
    }

    return floor(0x7FFFFFFFl / maxDouble);
}

PWIZ_API_DECL
size_t encodeLinear(
        const double *data, 
        size_t dataSize, 
        unsigned char *result,
        double fixedPoint
) {
    unsigned long long ints[3];
    size_t i, ri;
    unsigned char halfBytes[10];
    size_t halfByteCount;
    size_t hbi;
    long long extrapol;
    int diff;

	if (!fixedPoint)
		fixedPoint = optimalLinearFixedPoint(data, dataSize); // not specified, figure it out - bpratt

    //printf("Encoding %d doubles with fixed point %f\n", (int)dataSize, fixedPoint);
    encodeFixedPoint(fixedPoint, result);


    if (dataSize == 0) return 8;

    ints[1] = data[0] * fixedPoint + 0.5;
    for (i=0; i<4; i++) {
        result[8+i] = (ints[1] >> (i*8)) & 0xff;
    }

    if (dataSize == 1) return 12;

    ints[2] = data[1] * fixedPoint + 0.5;
    for (i=0; i<4; i++) {
        result[12+i] = (ints[2] >> (i*8)) & 0xff;
    }

    halfByteCount = 0;
    ri = 16;

    for (i=2; i<dataSize; i++) {
        ints[0] = ints[1];
        ints[1] = ints[2];
        ints[2] = data[i] * fixedPoint + 0.5;
        extrapol = ints[1] + (ints[1] - ints[0]);
        diff = ints[2] - extrapol;
        //printf("%lu %lu %lu,   extrapol: %ld    diff: %d \n", ints[0], ints[1], ints[2], extrapol, diff);
        encodeInt(diff, &halfBytes[halfByteCount], &halfByteCount);
        /*
        printf("%d (%d):  ", diff, (int)halfByteCount);
        for (size_t j=0; j<halfByteCount; j++) {
            printf("%x ", halfBytes[j] & 0xf);
        }
        printf("\n");
        */
        
        
        for (hbi=1; hbi < halfByteCount; hbi+=2) {
            result[ri] = (halfBytes[hbi-1] << 4) | (halfBytes[hbi] & 0xf);
            //printf("%x \n", result[ri]);
            ri++;
        }
        if (halfByteCount % 2 != 0) {
            halfBytes[0] = halfBytes[halfByteCount-1];
            halfByteCount = 1;
        } else {
            halfByteCount = 0;
        }
    }
    if (halfByteCount == 1) {
        result[ri] = halfBytes[0] << 4;
        ri++;
    }
    return ri;
}


PWIZ_API_DECL
size_t decodeLinear(
        const unsigned char *data,
        const size_t dataSize,
        double *result
) {
    size_t i;
    size_t ri = 0;
    unsigned int init;
    int diff;
    long long ints[3];
    //double d;
    size_t di;
    int half;
    long long extrapol;
    long long y;
    double fixedPoint;
    
    //printf("Decoding %d bytes with fixed point %f\n", (int)dataSize, fixedPoint);

    if (dataSize < 8) return -1;
    
    fixedPoint = decodeFixedPoint(data);


    if (dataSize < 12) return -1;
    
    try {
        ints[1] = 0;
        for (i=0; i<4; i++) {
            ints[1] = ints[1] | ((0xff & (init = data[8+i])) << (i*8));
        }
        result[0] = ints[1] / fixedPoint;

        if (dataSize == 12) return 1;
        if (dataSize < 16) return -1;

        ints[2] = 0;
        for (i=0; i<4; i++) {
            ints[2] = ints[2] | ((0xff & (init = data[12+i])) << (i*8));
        }
        result[1] = ints[2] / fixedPoint;
            
        half = 0;
        ri = 2;
        di = 16;
        
        while (di < dataSize) {
            ints[0] = ints[1];
            ints[1] = ints[2];
            if (di == (dataSize - 1) && half == 1) {
                if ((data[di] & 0xf) != 0x8) {
                    break;
                }
            }
            decodeInt(data, &di, &half, &diff);
            
            extrapol = ints[1] + (ints[1] - ints[0]);
            y = extrapol + diff;
            //printf("%lu %lu,   extrapol: %ld    diff: %d \n", ints[0], ints[1], extrapol, diff);
            result[ri++]     = y / fixedPoint;
            ints[2]         = y;
        }
    } catch (...) {
        cerr << "DECODE ERROR" << endl;
        cerr << "i: " << i << endl;
        cerr << "ri: " << ri << endl;
        cerr << "di: " << di << endl;
        cerr << "half: " << half << endl;
        cerr << "dataSize: " << dataSize << endl;
        cerr << "ints[]: " << ints[0] << ", " << ints[1] << ", " << ints[2] << endl;
        cerr << "extrapol: " << extrapol << endl;
        cerr << "y: " << y << endl;

        for (i = di - 3; i < min(di + 3, dataSize); i++) {
            cerr << "data[" << i << "] = " << data[i];
        }
        cerr << endl;
    }
    
    return ri;
}

PWIZ_API_DECL
void encodeLinear(
        const std::vector<double> &data, 
        std::vector<unsigned char> &result,
        double fixedPoint
) {
    size_t dataSize = data.size();
    result.resize(dataSize * 5);
    size_t encodedLength = encodeLinear(&data[0], dataSize, &result[0], fixedPoint);
    result.resize(encodedLength);
}

PWIZ_API_DECL
void decodeLinear(
        const std::vector<unsigned char> &data,
        std::vector<double> &result
) {
    size_t dataSize = data.size();
    result.resize(dataSize * 2);
    size_t decodedLength = decodeLinear(&data[0], dataSize, &result[0]);
    result.resize(decodedLength);
}

/////////////////////////////////////////////////////////////

PWIZ_API_DECL
size_t encodePic(
        const double *data, 
        size_t dataSize, 
        unsigned char *result
) {
    size_t i, ri, count;
    unsigned char halfBytes[10];
    size_t halfByteCount;
    size_t hbi;

    //printf("Encoding %d doubles\n", (int)dataSize);

    halfByteCount = 0;
    ri = 0;

    for (i=0; i<dataSize; i++) {
        count = data[i] + 0.5;
        //printf("%d %d %d,   extrapol: %d    diff: %d \n", ints[0], ints[1], ints[2], extrapol, diff);
        encodeInt(count, &halfBytes[halfByteCount], &halfByteCount);
        /*
        printf("%d (%d):  ", count, (int)halfByteCount);
        for (j=0; j<halfByteCount; j++) {
            printf("%x ", halfBytes[j] & 0xf);
        }
        printf("\n");
        */
        
        for (hbi=1; hbi < halfByteCount; hbi+=2) {
            result[ri] = (halfBytes[hbi-1] << 4) | (halfBytes[hbi] & 0xf);
            //printf("%x \n", result[ri]);
            ri++;
        }
        if (halfByteCount % 2 != 0) {
            halfBytes[0] = halfBytes[halfByteCount-1];
            halfByteCount = 1;
        } else {
            halfByteCount = 0;
        }
    }
    if (halfByteCount == 1) {
        result[ri] = halfBytes[0] << 4;
        ri++;
    }
    return ri;
}




PWIZ_API_DECL
size_t decodePic(
        const unsigned char *data,
        const size_t dataSize,
        double *result
) {
    size_t i, ri;
    int count;
    //double d;
    size_t di;
    int half;

    try {
        half = 0;
        ri = 0;
        di = 0;
        
        while (di < dataSize) {
            if (di == (dataSize - 1) && half == 1) {
                if ((data[di] & 0xf) != 0x8) {
                    break;
                }
            }
            decodeInt(&data[0], &di, &half, &count);
            
            //printf("count: %d \n", count);
            result[ri++]     = count;
        }
    } catch (...) {
        cerr << "DECODE ERROR" << endl;
        cerr << "ri: " << ri << endl;
        cerr << "di: " << di << endl;
        cerr << "half: " << half << endl;
        cerr << "dataSize: " << dataSize << endl;
        cerr << "count: " << count << endl;

        for (i = di - 3; i < min(di + 3, dataSize); i++) {
            cerr << "data[" << i << "] = " << data[i];
        }
        cerr << endl;
    }
    return ri;
}


PWIZ_API_DECL
void encodePic(
        const std::vector<double> &data,  
        std::vector<unsigned char> &result
) {
    size_t dataSize = data.size();
    result.resize(dataSize * 5);
    size_t encodedLength = encodePic(&data[0], dataSize, &result[0]);
    result.resize(encodedLength);
}


PWIZ_API_DECL
void decodePic(
        const std::vector<unsigned char> &data,  
        std::vector<double> &result
) {
    size_t dataSize = data.size();
    result.resize(dataSize * 2);
    size_t decodedLength = decodePic(&data[0], dataSize, &result[0]);
    result.resize(decodedLength);
}

/////////////////////////////////////////////////////////////

PWIZ_API_DECL
double optimalSlofFixedPoint(
        const double *data, 
        size_t dataSize
) {
    if (dataSize == 0) return 0;
    
    double maxDouble = 1;
    double x;
    double fp;

    for (size_t i=0; i<dataSize; i++) {
        x = log(data[i]+1);
        maxDouble = max(maxDouble, x);
    }

    fp = floor(0xFFFF / maxDouble);

    //cout << "    max val: " << maxDouble << endl;
    //cout << "fixed point: " << fp << endl;

    return fp;
}

PWIZ_API_DECL
size_t encodeSlof(
        const double *data, 
        size_t dataSize, 
        unsigned char *result,
        double fixedPoint
) {
    size_t i, ri;
    unsigned short x;
    if (!fixedPoint) // not specified, so derive - bpratt
        fixedPoint=optimalSlofFixedPoint(data, dataSize);
    encodeFixedPoint(fixedPoint, result);

    ri = 8; // first 8 bytes hold the fixedpoint value
    for (i=0; i<dataSize; i++) {
        x = log(data[i]+1) * fixedPoint + 0.5;
        result[ri++] = x & 0xff;
        result[ri++] = x >> 8;
    }
    return ri;
}


PWIZ_API_DECL
size_t decodeSlof(
        const unsigned char *data, 
        const size_t dataSize, 
        double *result
) {
    size_t i, ri;
    unsigned short x;
    ri = 0;
    double fixedPoint;

    if (dataSize < 8) return -1;
    fixedPoint = decodeFixedPoint(data);

    for (i=8; i<dataSize; i+=2) {
        x = data[i] | (data[i+1] << 8);
        result[ri++] = exp(x / fixedPoint) - 1;
    }
    return ri;
}


PWIZ_API_DECL
void encodeSlof(
        const std::vector<double> &data,  
        std::vector<unsigned char> &result,
        double fixedPoint
) {
    size_t dataSize = data.size();
    result.resize(dataSize * 2);
    size_t encodedLength = encodeSlof(&data[0], dataSize, &result[0], fixedPoint);
    result.resize(encodedLength);
}


PWIZ_API_DECL
void decodeSlof(
        const std::vector<unsigned char> &data,  
        std::vector<double> &result
) {
    size_t dataSize = data.size();
    result.resize(dataSize / 2);
    size_t decodedLength = decodeSlof(&data[0], dataSize, &result[0]);
    result.resize(decodedLength);
}

} // namespace MSNumpress
} // namespace msdata
} // namespace pwiz

File: pwiz-sharp/pwiz/src/MsData/MsNumpress.cs
namespace Pwiz.Data.MsData.Encoding;

/// <summary>
/// Numerical compressions for mass-spectrum binary arrays.
/// Port of <c>pwiz::msdata::MSNumpress</c> (Johan Teleman, 2013).
/// </summary>
/// <remarks>
/// Three lossy codecs that produce smaller payloads than raw float/double:
/// <list type="bullet">
///   <item><see cref="EncodeLinear(double[], double)"/> — fixed-point + linear prediction (m/z, retention time).</item>
///   <item><see cref="EncodePic(double[])"/> — nearest-integer compression (ion counts).</item>
///   <item><see cref="EncodeSlof(double[], double)"/> — short-logged-float (ion counts, more dynamic range).</item>
/// </list>
/// Integer values are packed using <see cref="EncodeInt(int, byte[], ref int)"/>, which stores a 4-byte
/// int as a variable-length sequence of half-bytes (nibbles).
/// </remarks>
public static class MsNumpress
{
    // ---------- encodeFixedPoint / decodeFixedPoint ----------

    /// <summary>Writes the 8 bytes of <paramref name="fixedPoint"/> in big-endian order into <paramref name="result"/>.</summary>
    /// <remarks>
    /// The ms-numpress reference implementation writes the double in big-endian byte order on every
    /// platform (<c>MSNumpress.cpp</c>: <c>result[i] = fp[IS_LITTLE_ENDIAN ? (7-i) : i]</c>, which
    /// reverses the bytes on little-endian hosts). Earlier versions called this <c>is_big_endian()</c>
    /// — the upstream has since been renamed but the wire format is unchanged. Byte-exact interop
    /// with pwiz-encoded streams requires matching this behavior.
    /// </remarks>
    public static void EncodeFixedPoint(double fixedPoint, byte[] result)
    {
        ArgumentNullException.ThrowIfNull(result);
        System.Buffers.Binary.BinaryPrimitives.WriteDoubleBigEndian(result.AsSpan(0, 8), fixedPoint);
    }

    /// <summary>Reads the 8-byte big-endian fixed-point double from <paramref name="data"/>.</summary>
    public static double DecodeFixedPoint(ReadOnlySpan<byte> data)
        => System.Buffers.Binary.BinaryPrimitives.ReadDoubleBigEndian(data[..8]);

    // ---------- encodeInt / decodeInt ----------

    /// <summary>
    /// Packs <paramref name="x"/> (a 32-bit signed int) as a variable-length sequence of half-bytes into
    /// <paramref name="res"/>, advancing <paramref name="resLength"/> by the number of half-bytes written.
    /// </summary>
    /// <remarks>
    /// The first half-byte <c>c</c> encodes how many leading zero/one half-bytes the int has:
    /// <c>0..8</c> = leading zeros, <c>9..15</c> = leading ones. Then the remaining half-bytes follow in
    /// little-endian order. Writes at most 9 half-bytes.
    /// </remarks>
    public static void EncodeInt(int x, byte[] res, ref int resLength)
    {
        ArgumentNullException.ThrowIfNull(res);
        const uint mask = 0xf0000000u;
        uint ux = (uint)x;
        uint init = ux & mask;

        if (init == 0)
        {
            int l = 8;
            for (int i = 0; i < 8; i++)
            {
                uint m = mask >> (4 * i);
                if ((ux & m) != 0) { l = i; break; }
            }
            res[resLength] = (byte)l;
            for (int i = l; i < 8; i++)
                res[resLength + 1 + i - l] = (byte)(ux >> (4 * (i - l)));
            resLength += 1 + 8 - l;
        }
        else if (init == mask)
        {
            int l = 7;
            for (int i = 0; i < 8; i++)
            {
                uint m = mask >> (4 * i);
                if ((ux & m) != m) { l = i; break; }
            }
            res[resLength] = (byte)(l + 8);
            for (int i = l; i < 8; i++)
                res[resLength + 1 + i - l] = (byte)(ux >> (4 * (i - l)));
            resLength += 1 + 8 - l;
        }
        else
        {
            res[resLength] = 0;
            for (int i = 0; i < 8; i++)
                res[resLength + 1 + i] = (byte)(ux >> (4 * i));
            resLength += 9;
        }
    }

    /// <summary>
    /// Decodes a single int from the nibble stream <paramref name="data"/>. <paramref name="di"/> is the
    /// current byte index; <paramref name="half"/> is 0 when we're reading the high nibble of that byte,
    /// 1 for the low nibble. Both are advanced as the int is consumed.
    /// </summary>
    public static void DecodeInt(ReadOnlySpan<byte> data, ref int di, ref int half, out int result)
    {
        byte head;
        if (half == 0) head = (byte)(data[di] >> 4);
        else { head = (byte)(data[di] & 0xf); di++; }

        half = 1 - half;
        uint res = 0;
        int n;
        if (head <= 8) n = head;
        else
        {
            n = head - 8;
            uint mask = 0xf0000000u;
            for (int i = 0; i < n; i++)
            {
                uint m = mask >> (4 * i);
                res |= m;
            }
        }

        if (n == 8) { result = (int)res; return; }

        for (int i = n; i < 8; i++)
        {
            byte hb;
            if (half == 0) hb = (byte)(data[di] >> 4);
            else { hb = (byte)(data[di] & 0xf); di++; }
            res |= (uint)hb << ((i - n) * 4);
            half = 1 - half;
        }
        result = (int)res;
    }

    // ---------- linear (m/z, retention time) ----------

    /// <summary>
    /// Returns the largest fixed-point multiplier that lets every linear-predicted residual fit into
    /// a 32-bit signed int. Used when a caller passes 0 for the fixed point.
    /// </summary>
    public static double OptimalLinearFixedPoint(ReadOnlySpan<double> data)
    {
        if (data.Length == 0) return 0;
        if (data.Length == 1) return Math.Floor(0x7FFFFFFFL / data[0]);
        double maxVal = Math.Max(data[0], data[1]);
        for (int i = 2; i < data.Length; i++)
        {
            double extrapol = data[i - 1] + (data[i - 1] - data[i - 2]);
            double diff = data[i] - extrapol;
            maxVal = Math.Max(maxVal, Math.Ceiling(Math.Abs(diff) + 1));
        }
        return Math.Floor(0x7FFFFFFFL / maxVal);
    }

    /// <summary>
    /// Returns a fixed-point multiplier that yields <paramref name="massAccuracy"/> absolute accuracy,
    /// or <c>-1</c> if that accuracy can't be achieved without 32-bit overflow.
    /// </summary>
    public static double OptimalLinearFixedPointMass(ReadOnlySpan<double> data, double massAccuracy)
    {
        if (data.Length < 3) return 0;
        double maxFp = 0.5 / massAccuracy;
        double overflow = OptimalLinearFixedPoint(data);
        return maxFp > overflow ? -1 : maxFp;
    }

    /// <summary>
    /// Encodes <paramref name="data"/> using fixed-point quantization plus linear-prediction residuals.
    /// When <paramref name="fixedPoint"/> is 0, <see cref="OptimalLinearFixedPoint"/> picks it.
    /// </summary>
    public static byte[] EncodeLinear(double[] data, double fixedPoint)
    {
        ArgumentNullException.ThrowIfNull(data);
        if (fixedPoint == 0) fixedPoint = OptimalLinearFixedPoint(data);

        var buf = new byte[8 + data.Length * 5];
        EncodeFixedPoint(fixedPoint, buf);
        if (data.Length == 0) return Trim(buf, 8);

        long i1 = (long)(data[0] * fixedPoint + 0.5);
        for (int i = 0; i < 4; i++) buf[8 + i] = (byte)((i1 >> (i * 8)) & 0xff);
        if (data.Length == 1) return Trim(buf, 12);

        long i2 = (long)(data[1] * fixedPoint + 0.5);
        for (int i = 0; i < 4; i++) buf[12 + i] = (byte)((i2 >> (i * 8)) & 0xff);

        int halfByteCount = 0;
        int ri = 16;
        byte[] halfBytes = new byte[10];
        long prev1 = i1, prev2 = i2;

        for (int i = 2; i < data.Length; i++)
        {
            long cur = (long)(data[i] * fixedPoint + 0.5);
            long extrapol = prev2 + (prev2 - prev1);
            int diff = (int)(cur - extrapol);
            EncodeInt(diff, halfBytes, ref halfByteCount);

            for (int hbi = 1; hbi < halfByteCount; hbi += 2)
            {
                buf[ri++] = (byte)((halfBytes[hbi - 1] << 4) | (halfBytes[hbi] & 0xf));
            }
            if (halfByteCount % 2 != 0)
            {
                halfBytes[0] = halfBytes[halfByteCount - 1];
                halfByteCount = 1;
            }
            else halfByteCount = 0;

            prev1 = prev2;
            prev2 = cur;
        }
        if (halfByteCount == 1)
        {
            buf[ri++] = (byte)(halfBytes[0] << 4);
        }
        return Trim(buf, ri);
    }

    /// <summary>Decodes a byte array produced by <see cref="EncodeLinear(double[], double)"/>.</summary>
    public static double[] DecodeLinear(ReadOnlySpan<byte> data)
    {
        if (data.Length < 8) throw new InvalidDataException("Numpress linear: need at least 8 bytes for fixed point.");
        double fixedPoint = DecodeFixedPoint(data);
        if (data.Length < 12) return Array.Empty<double>();

        var result = new double[(data.Length - 8) * 2];
        long i1 = 0;
        for (int i = 0; i < 4; i++) i1 |= (long)(data[8 + i] & 0xff) << (i * 8);
        // sign-extend from 32 bits
        i1 = (int)i1;
        result[0] = i1 / fixedPoint;
        if (data.Length == 12) return Resize(result, 1);
        if (data.Length < 16) throw new InvalidDataException("Numpress linear: truncated data after first value.");

        long i2 = 0;
        for (int i = 0; i < 4; i++) i2 |= (long)(data[12 + i] & 0xff) << (i * 8);
        i2 = (int)i2;
        result[1] = i2 / fixedPoint;

        int ri = 2;
        int di = 16;
        int half = 0;
        long prev1 = i1, prev2 = i2;

        while (di < data.Length)
        {
            if (di == data.Length - 1 && half == 1)
            {
                if ((data[di] & 0xf) != 0x8) break;
            }
            DecodeInt(data, ref di, ref half, out int diff);
            long extrapol = prev2 + (prev2 - prev1);
            long y = extrapol + diff;
            result[ri++] = y / fixedPoint;
            prev1 = prev2;
            prev2 = y;
        }
        return Resize(result, ri);
    }

    // ---------- pic (ion counts, integer) ----------

    /// <summary>
    /// Encodes <paramref name="data"/> as nearest-integer counts compressed with <see cref="EncodeInt(int, byte[], ref int)"/>.
    /// Handleable value range: <c>0 .. 4294967294</c>.
    /// </summary>
    public static byte[] EncodePic(double[] data)
    {
        ArgumentNullException.ThrowIfNull(data);
        var buf = new byte[data.Length * 5 + 1]; // +1 slack for odd half-byte trailer
        int halfByteCount = 0;
        int ri = 0;
        byte[] halfBytes = new byte[10];

        for (int i = 0; i < data.Length; i++)
        {
            int count = (int)(data[i] + 0.5);
            EncodeInt(count, halfBytes, ref halfByteCount);
            for (int hbi = 1; hbi < halfByteCount; hbi += 2)
            {
                buf[ri++] = (byte)((halfBytes[hbi - 1] << 4) | (halfBytes[hbi] & 0xf));
            }
            if (halfByteCount % 2 != 0)
            {
                halfBytes[0] = halfBytes[halfByteCount - 1];
                halfByteCount = 1;
            }
            else halfByteCount = 0;
        }
        if (halfByteCount == 1)
        {
            buf[ri++] = (byte)(halfBytes[0] << 4);
        }
        return Trim(buf, ri);
    }

    /// <summary>Decodes a byte array produced by <see cref="EncodePic"/>.</summary>
    public static double[] DecodePic(ReadOnlySpan<byte> data)
    {
        var result = new double[data.Length * 2];
        int ri = 0;
        int di = 0;
        int half = 0;
        while (di < data.Length)
        {
            if (di == data.Length - 1 && half == 1)
            {
                if ((data[di] & 0xf) != 0x8) break;
            }
            DecodeInt(data, ref di, ref half, out int count);
            result[ri++] = count;
        }
        return Resize(result, ri);
    }

    // ---------- slof (short-logged-float, ion counts) ----------

    /// <summary>Returns the fixed-point multiplier that maximizes Slof precision for this data.</summary>
    public static double OptimalSlofFixedPoint(ReadOnlySpan<double> data)
    {
        if (data.Length == 0) return 0;
        double maxVal = 1;
        for (int i = 0; i < data.Length; i++)
            maxVal = Math.Max(maxVal, Math.Log(data[i] + 1));
        return Math.Floor(0xFFFF / maxVal);
    }

    /// <summary>
    /// Encodes <paramref name="data"/> as <c>ushort fp = round(log(x+1) * fixedPoint)</c>.
    /// When <paramref name="fixedPoint"/> is 0, <see cref="OptimalSlofFixedPoint"/> picks it.
    /// </summary>
    public static byte[] EncodeSlof(double[] data, double fixedPoint)
    {
        ArgumentNullException.ThrowIfNull(data);
        if (fixedPoint == 0) fixedPoint = OptimalSlofFixedPoint(data);

        var buf = new byte[8 + data.Length * 2];
        EncodeFixedPoint(fixedPoint, buf);
        int ri = 8;
        for (int i = 0; i < data.Length; i++)
        {
            ushort x = (ushort)(Math.Log(data[i] + 1) * fixedPoint + 0.5);
            buf[ri++] = (byte)(x & 0xff);
            buf[ri++] = (byte)(x >> 8);
        }
        return buf;
    }

    /// <summary>Decodes a byte array produced by <see cref="EncodeSlof"/>.</summary>
    public static double[] DecodeSlof(ReadOnlySpan<byte> data)
    {
        if (data.Length < 8) throw new InvalidDataException("Numpress slof: need at least 8 bytes for fixed point.");
        double fixedPoint = DecodeFixedPoint(data);
        int count = (data.Length - 8) / 2;
        var result = new double[count];
        for (int i = 0; i < count; i++)
        {
            ushort x = (ushort)(data[8 + i * 2] | (data[8 + i * 2 + 1] << 8));
            result[i] = Math.Exp(x / fixedPoint) - 1;
        }
        return result;
    }

    // ---------- helpers ----------

    private static byte[] Trim(byte[] buf, int length)
    {
        if (buf.Length == length) return buf;
        var copy = new byte[length];
        Array.Copy(buf, copy, length);
        return copy;
    }

    private static double[] Resize(double[] arr, int length)
    {
        if (arr.Length == length) return arr;
        var copy = new double[length];
        Array.Copy(arr, copy, length);
        return copy;
    }
}
